### ISSUE TITLE
failed to find path /install/lib/libpython3.12.dylib

### ISSUE THREAD

**Issue by my1e5** [Nuitka/Nuitka#2933]
- Nuitka version, full Python version, flavor, OS, etc. as output by *this exact* command.
```
❯ python -m nuitka --version
2.3.4
Commercial: None
Python: 3.12.3 (main, Apr 15 2024, 17:43:11) [Clang 17.0.6 ]
Flavor: Unknown
Executable: /Users/rd/Documents/Users/my1e5/hello-nuitka/.venv/bin/python
OS: Darwin
Arch: arm64
macOSRelease: 14.5
Version C compiler: /usr/bin/clang (clang 15.0.0).
```
- How did you install Nuitka and Python

I'm using `rye` (https://github.com/astral-sh/rye) to manage my Python projects.

`rye add nuitka` adds `nuitka` as a dependency to project and installs it the `.venv`

- Also supply a Short, Self Contained, Correct, Example

Create a new project using `rye` and add `nuitka` as a dependency.
```
$ rye init --script hello-nuitka
$ cd hello-nuitka/
$ rye add nuitka
Added nuitka>=2.3.4 as regular dependency
Reusing already existing virtualenv
Generating production lockfile: /Users/rd/Documents/Users/my1e5/hello-nuitka/requirements.lock
Generating dev lockfile: /Users/rd/Documents/Users/my1e5/hello-nuitka/requirements-dev.lock
Installing dependencies
   Built file:///Users/rd/Documents/Users/my1e5/hello-nuitka                   Built 1 editable in 303ms
Installed 4 packages in 9ms
 + hello-nuitka==0.1.0 (from file:///Users/rd/Documents/Users/my1e5/hello-nuitka)
 + nuitka==2.3.4
 + ordered-set==4.1.0
 + zstandard==0.22.0
Done!
```
add a file in the root directory called `foo.py`
```
# foo.py
print("hello world")
```
```
$ python foo.py
hello world
```
Using nuitka with `--standalone`
```
❯ python -m nuitka --standalone foo.py 
Nuitka-Options: Used command line options: --standalone foo.py
Nuitka-Options: Detected static libpython to exist, consider '--static-libpython=yes' for better performance, but errors may happen.
Nuitka: Starting Python compilation with Nuitka '2.3.4' on Python '3.12' commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.                    
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka-Scons: Backend C compiler: clang (clang 15.0.0).
Nuitka-Scons: Backend linking program with 6 files (no progress information available for this stage).
Nuitka-Scons: Compiled 6 C files using ccache.
Nuitka-Scons: Cached C files (using ccache) with result 'cache hit': 5
Nuitka-Scons: Cached C files (using ccache) with result 'cache miss': 1
FATAL: Error, failed to find path /install/lib/libpython3.12.dylib (resolved DLL to /install/lib/libpython3.12.dylib) for foo.dist/foo.bin from 'None', please report the bug.
```
Without standalone
```
❯ python -m nuitka foo.py 
Nuitka-Options: Used command line options: foo.py
Nuitka-Options:WARNING: You did not specify to follow or include anything but main program. Check options and make sure that is intended.
Nuitka-Options: Detected static libpython to exist, consider '--static-libpython=yes' for better performance, but errors may happen.
Nuitka: Starting Python compilation with Nuitka '2.3.4' on Python '3.12' commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka-Scons: Backend C compiler: clang (clang 15.0.0).
Nuitka-Scons: Backend linking program with 6 files (no progress information available for this stage).
Nuitka-Scons: Compiled 6 C files using ccache.
Nuitka-Scons: Cached C files (using ccache) with result 'cache hit': 4
Nuitka-Scons: Cached C files (using ccache) with result 'cache miss': 2
Nuitka: Keeping build directory 'foo.build'.
Nuitka: Successfully created 'foo.bin'.
```
```
❯ ./foo.bin
dyld[43992]: Library not loaded: /install/lib/libpython3.12.dylib
  Referenced from: <BD10CCF2-7643-3ADE-8287-89E2EDB7660A> /Users/rd/Documents/Users/my1e5/hello-nuitka/foo.bin
  Reason: tried: '/install/lib/libpython3.12.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/install/lib/libpython3.12.dylib' (no such file), '/install/lib/libpython3.12.dylib' (no such file)
zsh: abort      ./foo.bin
```

**Comment by kayhayen** [comment]
I don't know what rye is, and your Python flavor being undetected is not a good sign. What is it?

**Comment by my1e5** [comment]
`rye` is a Python packaging and dependency manager similar to `pdm` and `poetry`. It's backed by the same developers making `ruff` - https://github.com/astral-sh/ruff.

`rye` manages the Python installation itself. Quoting from https://rye.astral.sh/guide/toolchains/

> Rye is unique in that it does not use system Python installations. Instead it downloads and manages Python installations itself (called toolchains). Today there are three types of toolchains supported by Rye and they require some understanding:
> 
> * [Portable CPython](https://rye.astral.sh/guide/toolchains/cpython/): Rye will itself download portable builds of CPython for most of its needs. These are fetched from [indygreg/python-build-standalone](https://github.com/indygreg/python-build-standalone)
> * [Official PyPy Builds](https://rye.astral.sh/guide/toolchains/pypy/): PyPy is supported from the official release builds.
> * [Custom Local Toolchains](https://rye.astral.sh/guide/toolchains/#registering-toolchains): locally installed Python interpreters can be registered with Rye. Afterwards, they can be used with any Rye managed project.

In my case I'm using the first option 'Portable CPython' which I believe is the default.

```
❯ rye toolchain list
cpython@3.12.3 (/Users/rd/.rye/py/cpython@3.12.3/bin/python3)
cpython@3.12.2 (/Users/rd/.rye/py/cpython@3.12.2/bin/python3)
```

**Comment by my1e5** [comment]
I've just seen that `It is known that macOS "pyenv" does not work.` Perhaps this is similar situation?

**Comment by kayhayen** [comment]
Well, supported is only what is listed as supported. I think pyenv is somehow misguided, but in principle I strive to support all Python versions.

The dependency on `libpython3.12.dylib` has to be adapted when used like that.

```
    # On macOS, we update the executable path for searching the "libpython"
    # library.
    if (
        isMacOS()
        and not Options.shallMakeModule()
        and not Options.shallUseStaticLibPython()
    ):
        python_abi_version = python_version_str + getPythonABI()
        python_dll_filename = "libpython" + python_abi_version + ".dylib"
        python_lib_path = os.path.join(sys.prefix, "lib")
        python_dll_path = os.path.join(python_lib_path, python_dll_filename)

        # Note: For CPython, and potentially others, the rpath for the Python
        # library needs to be set, so it will be detected as a dependency
        # without tricks.
        callInstallNameTool(
            filename=result_filename,
            mapping=(
                (
                    python_dll_filename,
                    python_dll_path,
                ),
                (
                    "@rpath/Python3.framework/Versions/%s/Python3" % python_version_str,
                    python_dll_path,
                ),
            ),
            id_path=None,
            rpath=python_lib_path,
        )
```

This code is used with CPython and maybe Anaconda and hardcodes the DLL dependency to replace. I guess all I need to do is make the input detected rather than hardcoded to the CPython scheme, and it will start working.

**Comment by kayhayen** [comment]
On current factory, you should now having it manage the libpython dependency.

**Comment by kayhayen** [comment]
Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by my1e5** [comment]
Just tried the factory branch and it works without any errors. Thanks!

**Comment by kayhayen** [comment]
Ok, that said, using `libpython.dylib` is a bad sign, Nuitka will be a lot better if you use the static libpython. Now I don't know how bad your flavor is there, so I tried, and it seems they cannot do LTO with static-libpython, so it's bad. They used LLVM 15, while my machine has LLVM 17, and they are not compatible, so that's not working. First time I see this, CPython does not exhibit this, but maybe it has no bytecode libpython.a, I wouldn't know. Disabling LTO in Nuitka also doesn't work, since this is a bytecode only libpython that my clang from Apple won't understand, maybe one can install their toolchain and then it works better?

**Comment by kayhayen** [comment]
This is currently on develop branch and will be in 2.4 release.

**Comment by kayhayen** [comment]
Yeah, them using `-flto=thin` is not good for compatibility with other toolkits.

**Comment by kayhayen** [comment]
Actually, it seems I am using clang 15, and they are using clang 17, so I don't know that, for now it might make sense to add a version check in Scons, so we error out more explicitly, but I think it's over the edge, so I added a detection of rye python on macOS and it won't suggest to use the static link library anymore.

**Comment by kayhayen** [comment]
Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by kayhayen** [comment]
Also one note, CPython just doesn't have a static link library. I need to revisit Anaconda on macOS, maybe it's better there with a conda package installed, at least on Linux that works as far as I recall. The best approach will be a yet to produce provided instruction by Nuitka project itself for how to build your Python. The rye Python build looks pretty good aside the thin LTO which prevents it from being used, but I guess, if you were to build locally, the clang would be the same it would work. Somebody else can look into this. 

I am closing this early for lack of general relevance. If somebody wants to feedback rye python about providing maybe a more suitable libpython or a way to use their tool chain (rye install clang installs a python package) that would be great. At least rye indicates probably that it's not pip, but `uf` in installer, I like that.

**Comment by kayhayen** [comment]
I found the build quality of rye somewhat lacking, e.g. the build is not reproducible, which is bad. I think rust people don't know or care to produce usable libraries. I have had plenty of issues with rust bindings producing crap output. This is not C after all, so their experience is probably non-existant. I am more a fan of zig. 

It seems to me people want to find tools to re-implement in rust, because they want to code in rust, rather than solving issues the tools have. I mean, pip is fast these days, and if you use pdm and the like, using those modern PEPs, it's not even like there is anything to do really.

But let me work on Nuitka scalability, or else somebody is going to rewrite it in in rust. ;-)

**Comment by bersbersbers** [comment]
> pip is fast these days

I wonder if you still say that after trying `uv pip` once, especially if you ignore download times in that comparison 😁 but other than that, I believe you are right: these tools are not solving "new" problems (unless you consider slowness of the existing tools a problem in itself).

**Comment by kayhayen** [comment]
The detection part made it into 2.3 hotfix release, the generic detection of dynamic libpython will come with 2.4 only.

**Comment by kayhayen** [comment]
Of course, I tried it only to see if it behaved well. Behaving well in my book means not being incompatible with Pip in the spirit of let's cut corners. I saw that happen before. Setting the `INSTALLER` file is one example of where that could have happened. There probably is more, but luckily, Nuitka doesn't care as much about bugs caused by wrong installation since they already affect Python, and it's only to replicate those, so with that file, I was already happy. 

I am not sure what difference I should have noticed. Also, I prefer to use tools like pipenv, which tends to realize that there is nothing to do, to begin with by caching the virtualenv. Direct usage of pip is just when I develop with Nuitka and people who used pip, that would be prudent, and with people using uv, that would be prudent. My tool to install from compilation reports aims to reproduce what people did, conda, Pacman, pip, and well, maybe uv now. I care about being bit-identical rather than saving, say, one minute at installation and wasting a work day of mine on not reproducing an issue. For me that is easily 2% less development for Nuitka this year, every time it happens.

Is there proof anywhere that the pip and uv are giving bit identical virtual environments? Pip has so many modes for legacy that I guess a new implementation would skip over them first.
